### DISCUSSION TITLE
How do I create a Class Library of User Controls?

### DISCUSSION THREAD

**Discussion by MarkBrainfire** [AvaloniaUI/Avalonia#12739] [General]
I've read the documentation here:
https://docs.avaloniaui.net/docs/next/guides/custom-controls/how-to-create-a-custom-controls-library
But I still don't understand.

Is the thinking that all User Controls should be stand alone projects?

The Avalonia UI templates only have two types of projects: an Avalonia C# application that defines a Window and an Avalonia User Control.  I'm used to also having a Class Library project type that can contain multiple User Controls and Classes and then be referenced by other Applications or Class Libraries providing access to all of the Controls and Classes defined within.

Obviously, I'm new and trying to figure the Avalonia approach out.  Why didn't you just include a Class Library as a project type?

**Comment by workgroupengineering** [comment]
see https://github.com/AvaloniaUI/Avalonia.Labs

**Comment by timunie** [comment]
UserControls (page content) or CustomControls? 

Custom, reusalbe controls can go into any library project. Only thing needed is to reference Avalonia version 11.0.0 (yes, 11.0.0, minor updates come from referencing project / unless you need a newer function for sure)

**Reply by MarkBrainfire** [reply]
I'm coming from the WinUI 3 world.  UserControls were relatively easy to put together from other existing controls or UserControls, but had dependency properties so you could create XAML attributes for them.  Custom controls were more complex but could be templated.

A "Page" was just that a visual Page that could be content for a Window.  Pages appear to be non-existent in Avalonia, which is different.  In WinUI 3 a Page might contain a UserControl, but you would never include a Page as a "control", a Page was always root content for a Window.

So I'm just trying to understand the Avalonia differences.

What I get at this point is that to create Class Library of UserControls you start with a regular .NET core class library and then add some Avalonia Nuget packages ... although I"m not sure which packages?

**Reply by timunie** [reply]
UserControl can be both in Avalonia, a Page or a set of some other controls. Properties can be defined in addition. See: 
- https://docs.avaloniaui.net/docs/next/guides/custom-controls/types-of-control 
- https://docs.avaloniaui.net/docs/next/guides/custom-controls/defining-properties
- https://github.com/AvaloniaUI/Avalonia.Samples#%EF%B8%8F-custom-controls-samples

Also check out this for other control libs and samples: https://github.com/AvaloniaCommunity/awesome-avalonia
